### PR TITLE
Fix for declaring test case in data provider

### PR DESCRIPTION
**Pull request by kubawerlos**
For https://github.com/sebastianbergmann/phpunit/issues/2833.

I cannot reproduce the problem from https://github.com/sebastianbergmann/phpunit/issues/2859 - as I suspected the minimal, self-contained, reproducing test case from https://github.com/sebastianbergmann/phpunit/issues/2859#issuecomment-344278138 is not triggering the error. _Obviously I have tested my original solution in project with over dozen of data providers and all went fine_.

@sebastianbergmann can you create a branch for the issue and merge this PR on it so @eddmann, @moorscode or @johnbillion will provide a test case for the problem?

### CODE REVIEW

File: tests/_files/DataProviderDeclaringClass/SecondTest.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Foo\DataProviderDeclaringClass;
+
+use PHPUnit\Framework\TestCase;
+
+class SecondTest extends TestCase
+{
+    const DUMMY = 'dummy';
+
+    public function testSecond()
```
Comment (gavinlove) [inline tests/_files/DataProviderDeclaringClass/SecondTest.php]: Change this to be a data provider and it is likely you will see the problem.

**Comment by kubawerlos** [comment]
@gavinlove like this https://github.com/sebastianbergmann/phpunit/pull/2861/commits/dcace3b5b8b71ff559c2ac74f4d4cbfc6478847a ? Of course I tried something like this yesterday, give me some credit - I tried to follow instructions from the https://github.com/sebastianbergmann/phpunit/issues/2859 - none of them work. More, few minutes ago I have run PHPUnit 5.7.24 on master of [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) which have 495 data providers an (with PHP 7.1.11-1+ubuntu16.04.1+deb.sury.org+1) - all went fine.

TBH it feels little bit irritated - I have discovered the bug - this one that test class loaded during data provider execution is not discovered - provided decent [test case](https://github.com/sebastianbergmann/phpunit/commit/7e8e2625c289a63285625dcdd657b824797a3717) to cover the problem and then made a fix. I have tried the fix with multiple projects I have on my PC and had no error. I'd say I proceeded with the bug quite well.

Then https://github.com/sebastianbergmann/phpunit/issues/2859 was reported with non-working reproductions steps (few snippets, none of them leading to the error). Or maybe I'm wrong here, @sebastianbergmann did you manage to reproduce it, so we can create PHPT to cover it? If not I'd like to fix for this issue be accepted and merged and https://github.com/sebastianbergmann/phpunit/issues/2859 be fixed not by reverting fix of the other bug, but in proper way - PHPT case and then fix (which BTW I'd be happy to implement).

**Comment by gavinlove** [comment]
@kubawerlos Your right the two providers is a not the right way to reproduce it. I looked at @EddMann said I was able to reproduce this.

I used your `FirstTest` class but I then changed my `phpunit.xml.dist` to the following

```
<?xml version="1.0" encoding="UTF-8"?>

<phpunit bootstrap="vendor/autoload.php">
    <testsuites>
        <testsuite name="Test Suite">
            <directory>./tests/</directory>
            <directory>./tests/*/</directory>
        </testsuite>
    </testsuites>
</phpunit>
```

Note that the same directory is listed twice here which  @eddmann mentioned yesterday. I suspect what is happening is that the directory is parsed twice and then later on phpunit filters it down to a single test.

**Comment by johnbillion** [comment]
Please don't be irritated! We're simply reporting what we're seeing.

AFAICT, the minimal test case I reported in #2859 is enough to reproduce the issue, but there must be something else which causes this, but I don't know what.

I don't think the issue is dependent on PHP version. The issue appears for me when running PHP 7.1 and 7.2.

If it helps, here's a Travis CI job that failed due to this fatal error: https://travis-ci.org/johnbillion/query-monitor/jobs/301942375 .

**Comment by gavinlove** [comment]
@kubawerlos @johnbillion I have pushed up a repo which reproduces it for me, which will hopefully help.

https://github.com/gavinlove/phpunit-bug-2859

**Comment by kubawerlos** [comment]
@gavinlove thanks, your first comment gave me a hint and I just uploaded test case for the problem. Fix in progress.

**Comment by kubawerlos** [comment]
@sebastianbergmann fix might looks as really dummy, but in fact `PHPUnit_Framework_TestSuite_DataProvider` is the only PHPUnit class that implements `PHPUnit_Framework_Test` and would wrongly go into the calling of `addTestSuite` 16 lines below the added `if`.

File: src/Framework/TestSuite.php [hunk 2]
```diff
@@ -87,6 +87,11 @@ class PHPUnit_Framework_TestSuite implements PHPUnit_Framework_Test, PHPUnit_Fra
      */
     private $iteratorFilter = null;
 
+    /**
+     * @var array
```
Comment (keradus) [inline src/Framework/TestSuite.php]: array of ? strings? then `string[]`

**Pull request merged**